src/main/java/graphql/nadel/engine/SourceQueryTransformer.java

```diff
@@ -0,0 +1,124 @@
+package graphql.nadel.engine;
+
+import graphql.execution.ExecutionContext;
+import graphql.execution.ExecutionStepInfo;
+import graphql.execution.ExecutionStepInfoFactory;
+import graphql.execution.MergedField;
+import graphql.language.Document;
+import graphql.language.Field;
+import graphql.language.FieldDefinition;
+import graphql.language.OperationDefinition;
+import graphql.language.Selection;
+import graphql.language.SelectionSet;
+import graphql.nadel.dsl.FieldDefinitionWithTransformation;
+import graphql.nadel.dsl.FieldTransformation;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static graphql.execution.MergedField.newMergedField;
+import static graphql.language.OperationDefinition.newOperationDefinition;
+import static graphql.language.SelectionSet.newSelectionSet;
+import static java.util.Collections.singletonList;
+
+public class SourceQueryTransformer {
+    private final ExecutionStepInfoFactory executionStepInfoFactory = new ExecutionStepInfoFactory();
+    private OperationDefinition.Builder operationDefinition = newOperationDefinition();
+
+    private SourceQueryTransformer() {
+    }
+
+    public static Document transform(List<MergedField> fields, ExecutionContext executionContext, ExecutionStepInfo executionStepInfo) {
+        return new SourceQueryTransformer().transformInternal(fields, executionContext, executionStepInfo);
+    }
+
+    private Document transformInternal(List<MergedField> fields, ExecutionContext executionContext, ExecutionStepInfo executionStepInfo) {
```

Comment:
I feel that here we can do much better than just transforming to a target service query. We can also generate reverse transformation function for each field, such that when we apply it to the result from the delegate (Map<String,Object>) we will get a ObjectExecutionResultNode. I feel that it will be the easiest way to do as:
- you will not have to traverse types again and apply reverse transformations in a separate class  (ie. currently in DelegatedResultToResultNode) 
- all the transformation logic will be in one place,
- the only thing that strategy will have to do will be to apply reverse function,  collect  Unresolved nodes (hydration fields) group by service, and execute them again

@andimarek what do you think?